doc/design/publish-1.md

```diff
@@ -0,0 +1,141 @@
+# Moving Parts
+
+This is a summary of system components that are relevant to publish and how they relate to eachother.
+
+### Dandiset
+A DANDI dataset.
+Each dandiset is guaranteed to have a draft Version, and can also have any number of published Versions.
+
+### Version
+Versions consist of a JSON blob of metadata and any number of Assets.
+There are two kinds of Version: draft and published.
+Published Versions are immutable, only draft Versions are editable (except admins, who have superpowers).
+Versions also track their validation status.
+
+### AssetBlob
+This is basically just a pointer to a blob in S3.
+AssetBlobs are central to how we implement deduplication of blobs in S3.
+Multiple Assets can refer to the same AssetBlob, meaning we only need to store each blob in S3 once.
+
+Whenever a new AssetBlob is created, a background task runs to compute the sha256 checksum (a "digest") of the blob.
+This can take a long time for large blobs, as it must stream all of the data from S3.
+
+### Asset
+Assets consist of a JSON blob of metadata and a reference to an AssetBlob.
+Versions and Assets are many-to-many; a Version can have any number of Assets, and an Asset can belong to any number of Versions (even, hypothetically, versions in different Dandisets).
+Assets also track their validation status.
+
+Similarly to Versions, there are two kinds of Asset: draft and published.
+Draft Assets are only allowed to belong to draft Versions.
+Published Assets are basically the same as draft Assets, but have some extra metadata included that was defined during the publish process (`datePublished`, `publishedBy`, etc).
+
+Note that published Assets can belong to both draft and published Versions.
+
+Uploading new data or modifying the metadata of an Asset doesn't actually modify the Asset.
+Instead, it creates a new Asset and replaces the reference to that Asset in the Version.
+This ensures that other Versions which might refer to the original Asset are unmodified, while the Version containing the Asset being modified reflects the changes correctly.
+
+When a published Asset is modified in this way, the publish metadata is stripped out, effectively creating a new draft Asset.
+
+_In general_, Assets should be immutable; any modification will instead create a new draft Asset.
+* One exception to this is calculating the AssetBlob digest.
+When the digest completes, it will be injected directly into the metadata of any Assets referencing that AssetBlob.
+* The other exception is creating a published Asset during the publish process.
+This involves directly injecting the new metadata fields into the draft Asset, effectively turning it into a published Asset in place.
```

Comment:
per my question and brief discussion, I do not think we should bother with mutating Assets in DB, and can keep the notion of draft vs published asset metadata just implemented while returning asset record from an endpoint for dandiset_version.  If `version != "draft"` (thus published) -- we populate fields relevant to Published version.  If `draft` - we don't. Then we would stay consistent with the notion of immutable "Asset metadata as provided by user" in DB.
The same goes for 

> When the digest completes, it will be injected directly into the metadata of any Assets referencing that AssetBlob.

and we do not inject anything into actual asset metadata record but just populate it "on the fly" from blob DB record.
And we avoid the need for "the publish metadata is stripped out" which might get hairy IMHO.

If there is a desire to use `contentUrl` as from the first published version for an asset, could be achieved at DB query level by searching asset in Versions with minimal non-draft version.  Or we just KISS it and provide version specific contentUrl regardless.